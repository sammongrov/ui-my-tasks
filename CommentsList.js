import React, { Component } from 'react';
import { View, TouchableOpacity, FlatList, BackHandler, Keyboard /* Alert */ } from 'react-native';
import PropTypes from 'prop-types';
import { styles } from 'react-native-theme';
import { Actions } from 'react-native-router-flux';
// import Swipeout from 'react-native-swipeout';
import AppUtil from '@mongrov/utils'; 
import _ from 'lodash';
import { Colors } from '@ui/theme_default';
import {
  NavBar,
  Text,
  Screen,
  Icon,
  Avatar,
  TextInput /* FontAwesomeIcon */,
} from '@ui/components';
// import DBManager from '../app/DBManager';
import {DBManager} from 'app-module';
import {Application} from '@mongrov/config';

export default class CommentsList extends Component {
  static propTypes = {
    title: PropTypes.string.isRequired,
  };

  static defaultProps = {};

  state = {
    commentText: '',
  };

  componentWillMount() {
    const { commentBoardId, commentCardId } = this.props;
    const fetchChecklist = {
      boardID: commentBoardId,
      cardID: commentCardId,
    };
    BackHandler.addEventListener('hardwareBackPress', this.handleBackPress);
    DBManager.cardComments.fetchComment(fetchChecklist);
  }

  componentWillUnmount = () => {
    const { commentCardId } = this.props;
    this._isMounted = false;
    BackHandler.removeEventListener('hardwareBackPress', this.handleBackPress);
    DBManager.cardComments.removeCommentsListner(commentCardId, this.updateComments);
  };

  componentDidMount = () => {
    const { commentCardId } = this.props;
    this._isMounted = true;
    DBManager.cardComments.addCommentsListner(commentCardId, this.updateComments);
  };

  updateComments = async () => {
    if (!this._isMounted) return;
    const { commentCardId } = this.props;
    const comments = DBManager.cardComments.findByCardIdAsList(commentCardId);
    if (comments) {
      const updateCommentList = Object.keys(comments).map((key) => comments[key]);
      const Comments = _.orderBy(updateCommentList, 'createdAt', 'desc');
      this.comments = Comments;
      this.setState({
        updateComments: Comments,
      });
    }
  };

  handleBackPress = () => {
    Actions.pop();
    return true;
  };

  fetchCommentTasks = () => {
    const { commentCardId, commentBoardId } = this.props;
    const fetchChecklist = {
      boardID: commentBoardId,
      cardID: commentCardId,
    };
    DBManager.cardComments.fetchComment(fetchChecklist);
  };

  addComments = (commentText) => {
    const { commentBoardId, commentCardId } = this.props;
    const { board } = DBManager._taskManager;
    const commentToSave = {
      title: commentText,
      boardId: commentBoardId,
      cardId: commentCardId,
      userId: DBManager.app.userId,
    };
    if (commentToSave) {
      board.createComment(commentToSave);
      this.fetchCommentTasks();
      Keyboard.dismiss();
    }
  };

  deleteComment = (comment) => {
    const { board } = DBManager._taskManager;
    const commentToDelete = {
      boardId: comment.boardId,
      cardId: comment.cardId,
      commentId: comment._id,
    };
    if (commentToDelete) {
      board.deleteComment(commentToDelete);
      this.fetchCommentTasks();
    }
  };

  keyExtractor = (item) => item._id;

  avatarComment = (avatar) => {
    if (avatar.avatarURL) {
      return (
        <Avatar
          statusColor="transparent"
          avatarUrl={avatar.avatarURL}
          avatarName={avatar.name}
          key={avatar.avatarURL}
          avatarSize={30}
        />
      );
    }
  };

  updateSize = (height) => {
    this.setState({
      height,
    });
  };

  renderRowComment = ({ item }) => {
    const id = item.userId;
    const user = DBManager.user.findById(id) || {};
    // const swipeBtns = [
    //   {
    //     component: (
    //       <View
    //         style={{
    //           flex: 1,
    //           alignItems: 'center',
    //           justifyContent: 'center',
    //         }}
    //       >
    //         <FontAwesomeIcon name="trash-o" type="material-community" color="#2E88FF" size={20} />
    //       </View>
    //     ),
    //     backgroundColor: '#FFF',
    //     underlayColor: 'rgba(0, 0, 0, 1, 0.6)',
    //     onPress: () => {
    //       Alert.alert(
    //         'Delete',
    //         'Do you want to delete Comment ?',
    //         [
    //           { text: 'No', onPress: () => {}, style: 'cancel' },
    //           {
    //             text: 'Yes',
    //             onPress: () => {
    //               this.deleteComment(item);
    //               this.fetchCommentTasks();
    //             },
    //           },
    //         ],
    //         { cancelable: false },
    //       );
    //     },
    //   },
    // ];
    return (
      // <Swipeout right={swipeBtns} autoClose={true} backgroundColor="transparent" buttonWidth={40}>
      <TouchableOpacity
        style={[styles.rowDirection, styles.paddingHorizontal10]}
        // onPress={() => {
        //   if (Actions.currentScene === 'Chat') {
        //     Actions.ChatRoom({ roomInfo: item });
        //   }
        // }}
      >
        <View style={{ paddingTop: 7 }}>{this.avatarComment(user)}</View>
        <View style={[styles.cListRowContainer]}>
          <View
            style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              alignItems: 'flex-end',
            }}
          >
            <View>
              <Text numberOfLines={1} style={styles.cListTitle}>
                {user.username}
              </Text>
            </View>
            <View>
              <Text
                numberOfLines={1}
                style={[styles.fontSize12, { color: Colors.TYP_GRAY }, { marginBottom: 5 }]}
              >
                {AppUtil.formatDate(item.createdAt).toString()}
              </Text>
            </View>
          </View>
          <View
            style={{
              flexDirection: 'row',
              flex: 1,
              justifyContent: 'space-between',
              alignItems: 'flex-start',
            }}
          >
            <View style={{ flexDirection: 'column', flex: 1, marginRight: 5 }}>
              <Text style={[styles.flex1, styles.cListLastMessage, { color: Colors.TYP_GRAY }]}>
                {item.text}
              </Text>
            </View>
          </View>
        </View>
      </TouchableOpacity>
      // </Swipeout>
    );
  };

  renderListcomment = () => {
    const { updateComments } = this.state;
    if (updateComments) {
      return (
        <FlatList
          keyExtractor={this.keyExtractor}
          data={updateComments}
          renderItem={this.renderRowComment}
        />
      );
    }
    return <View style={{ flex: 1 }} />;
  };

  render() {
    const { commentText, height } = this.state;
    const newStyle = { height };

    return (
      <Screen>
        <NavBar
          titleText="Comments"
          leftComponent={
            <TouchableOpacity
              style={[styles.navSideButtonDimension, styles.alignJustifyCenter]}
              onPress={() => {
                if (
                  Actions.currentScene === 'CommentsList' ||
                  Actions.currentScene === 'CommentsListScene'
                ) {
                  Actions.pop();
                }
              }}
            >
              <Icon
                name="chevron-left"
                type="material-community"
                color={Colors.NAV_ICON}
                size={36}
              />
            </TouchableOpacity>
          }
        />
        <View style={{ flex: 1, justifyContent: 'space-between' }}>
          {this.renderListcomment()}
          {Application.APPCONFIG.BOARD_EDIT_UPDATE && (
            <View
              style={{
                flexDirection: 'row',
                paddingHorizontal: 10,
                paddingVertical: 10,
                alignItems: 'center',
              }}
            >
              <TextInput
                placeholder="Write comments"
                style={[styles.flex1, styles.imagePreviewTextInput, newStyle]}
                multiline
                onChangeText={(text) => this.setState({ commentText: text })}
                onContentSizeChange={(e) => this.updateSize(e.nativeEvent.contentSize.height)}
                value={commentText}
                underlineColorAndroid="transparent"
              />
              <TouchableOpacity
                style={{
                  height: 40,
                  width: 40,
                  backgroundColor: '#2E88FF',
                  borderRadius: 50,
                  justifyContent: 'center',
                  alignItems: 'center',
                }}
                onPress={() => {
                  this.addComments(commentText);
                  this.setState({
                    commentText: '',
                  });
                  this.fetchCommentTasks();
                }}
              >
                <Icon name="send" size={20} color="#fff" />
              </TouchableOpacity>
            </View>
          )}
        </View>
      </Screen>
    );
  }
}

CommentsList.propTypes = {
  commentBoardId: PropTypes.string,
  commentCardId: PropTypes.string,
};

CommentsList.defaultProps = {
  commentBoardId: '',
  commentCardId: ',',
};
