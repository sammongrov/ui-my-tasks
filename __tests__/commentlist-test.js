import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Actions } from 'react-native-router-flux';
import { BackHandler } from 'react-native';
import DbManager from '../../app/DBManager';
import CommentsList from '../CommentsList';

configure({ adapter: new Adapter() });

jest.mock('BackHandler', () => {
  const backHandler = {
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
  };
  return backHandler;
});

jest.mock('react-native-router-flux', () => ({
  Actions: {
    pop: jest.fn(),
    currentScene: 'BoardTasksList',
    CardDetails: jest.fn(),
    cardId: jest.fn(),
  },
}));
jest.mock('../../app/DBManager', () => {
  const dbManager = {
    group: {
      addCommentsListner: jest.fn(),
      removeCommentsListner: jest.fn(),
      fetchComment: jest.fn(),
    },

    cardComments: {
      fetchComment: jest.fn(() => ({ cardID: 'DFJEFHE', boardID: 'TEUFNDK' })),
      addCommentsListner: jest.fn(() => ({ commentCardId: 'YEIW3J' })),
      removeCommentsListner: jest.fn(() => ({ commentCardId: 'YEIW3J' })),
    },
  };
  return dbManager;
});
beforeEach(() => {
  jest.resetModules();
});

it('CommentsList renders correctly', () => {
  const tree = renderer.create(<CommentsList />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('handleBackPress', () => {
  const rootComponent = shallow(<CommentsList />);
  const instance = rootComponent.instance();
  const result = instance.handleBackPress();

  expect(result).toBe(true);
  expect(Actions.pop.mock.calls.length).toBe(1);
});

it('ChatMessageList - componentWillMount', async () => {
  DbManager.cardComments.fetchComment = jest.fn(() => ({ value: 123456 }));
  const tree = shallow(<CommentsList />);
  const instance = tree.instance();
  await instance.componentWillMount();
  expect(DbManager.cardComments.fetchComment).toBeCalled();
  expect(BackHandler.addEventListener).toBeCalled();
});

it('ChatMessageList - componentWillUnmount', () => {
  const tree = shallow(<CommentsList />);
  const instance = tree.instance();
  tree.unmount();
  expect(instance._isMounted).toBe(false);
  expect(DbManager.cardComments.removeCommentsListner).toBeCalled();
  expect(BackHandler.removeEventListener).toBeCalled();
});

it('ReplyMessage componentDidMount', () => {
  DbManager.cardComments.addCommentsListner.mockClear();
  const tree = shallow(<CommentsList />);
  const instance = tree.instance();
  expect(DbManager.cardComments.addCommentsListner).toBeCalled();
  expect(instance._isMounted).toBe(true);
});
it('key extractorcurrent scene is "CommentsList"', () => {
  const tree = renderer.create(<CommentsList cardId="7X856YIks9" />);
  const treeInstance = tree.getInstance();
  treeInstance.keyExtractor('XPP89611P');
  expect(Actions.cardId).toBeCalled();
});

it('Navbar onPress calls Actions.pop', () => {
  Actions.currentScene = 'CommentsList';
  Actions.currentScene = 'CommentsListScene';

  const root = shallow(<CommentsList />);
  const button = root
    .find('NavBar')
    .shallow()
    .find('TouchableOpacity');
  button.props().onPress();
  expect(Actions.pop.mock.calls.length).toBe(2);
});
it('Navbar onPress not to call Actions.pop', () => {
  Actions.currentScene = 'Comments';
  Actions.currentScene = 'CommentScene';
  Actions.pop.mockClear();
  const root = shallow(<CommentsList />);
  const button = root
    .find('NavBar')
    .shallow()
    .find('TouchableOpacity');
  button.props().onPress();
  expect(Actions.pop.mock.calls.length).toBe(0);
});
it('updateSize', () => {
  const height = 720;
  const rootComponent = shallow(<CommentsList />);

  const instance = rootComponent.instance();
  instance.updateSize(height);
  expect(rootComponent.state().height).toBe(height);
});
describe('CommentsList calls avatarComment', () => {
  const avatar = {
    avatarUrl: ' https://www.pexels.com/photo/nature-red-forest-leaves-33109/',
  };

  it('avatar comment ', () => {
    const tree = shallow(<CommentsList />);
    const instance = tree.instance();
    const avatarUrl = instance.avatarComment(avatar);
    expect(avatarUrl).toBeUndefined();
  });
});
describe('CommentsList calls renderRowComment', () => {
  const item = {
    userId: 'YEOENDK',
  };

  it('avatar comment ', () => {
    const tree = shallow(<CommentsList />);
    const instance = tree.instance();
    const RowComment = instance.renderRowComment(item);
    expect(RowComment).toBeUndefined();
  });
});
describe('CommentsList calls renderRowComment', () => {
  const commentToDelete = {
    boardId: 'TWUH79JEUHW',
    cardId: 'YE68WEBKE',
    commentId: 'YW083HR8',
  };

  it('avatar comment ', () => {
    const tree = shallow(<CommentsList />);
    const instance = tree.instance();
    const Commentdelete = instance.deleteComment(commentToDelete);
    expect(Commentdelete).toBeUndefined();
    expect(commentToDelete).toBeUndefined();
  });
});
describe('CommentsList calls addComments', () => {
  const commentToSave = {
    title: 'commentText',
    boardId: 'RT572JWSK8',
    cardId: 'JK893FENF',
    userId: 'YURW0QW8J',
  };
  it('avatar comment ', () => {
    const tree = shallow(<CommentsList />);
    const instance = tree.instance();
    const commentText = instance.addComments();
    expect(commentText).toBeUndefined();
    expect(commentText).toMatchObject(commentToSave);
  });
});
describe('CommentsList calls renderRowComment', () => {
  const fetchChecklist = {
    boardId: 'TWUH79JEUHW',
    cardId: 'YE68WEBKE',
  };

  it('avatar comment ', () => {
    const tree = shallow(<CommentsList />);
    const instance = tree.instance();
    const commentText = instance.fetchCommentTasks(fetchChecklist);
    expect(commentText).toBeUndefined();
    expect(fetchChecklist).toBeUndefined();
  });
});
describe('CommentsList calls renderRowComment', () => {
  it('avatar comment ', () => {
    const tree = shallow(<CommentsList />);
    const instance = tree.instance();
    const commentText = instance.updateComments();
    expect(commentText).toBeUndefined();
  });
});
describe('CommentsList calls renderRowComment', () => {
  const { updateComments } = this.state;

  it('avatar comment ', () => {
    const tree = shallow(<CommentsList />);
    const instance = tree.instance();
    const commentText = instance.renderListcomment(updateComments);
    expect(commentText).toBeUndefined();
  });
});
it('should render the onChange to textName', () => {
  const rootComponent = shallow(<CommentsList />);
  const authorInputComponent = rootComponent.find('TextInput').first();
  authorInputComponent.props().onChangeText('text');
  expect(rootComponent.state('commentText')).toEqual('text');
});

it('calls onContentSizeChange', () => {
  const rootComponent = shallow(<CommentsList />);
  const event = { nativeEvent: { contentSize: { height: 720 } } };
  const input = rootComponent.find('TextInput').first();
  input.props().onContentSizeChange(event);
  expect(rootComponent.state('height')).toEqual(720);
});
