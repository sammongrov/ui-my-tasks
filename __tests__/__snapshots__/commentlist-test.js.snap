// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentsList renders correctly 1`] = `
<BVLinearGradient
  colors={
    Array [
      4293914607,
      4293914607,
    ]
  }
  locations={
    Array [
      0,
      1,
    ]
  }
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Array [
            undefined,
            null,
          ]
        }
      >
        <View
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#2E88FF",
                    "fontSize": 36,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
              type="material-community"
            >
              
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              undefined,
              undefined,
              null,
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              Array [
                Object {
                  "fontFamily": "OpenSans-Regular",
                },
                Array [
                  undefined,
                  null,
                ],
              ]
            }
          >
            Comments
          </Text>
        </View>
        <View
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        />
      </View>
      <View
        style={
          Object {
            "flex": 1,
            "justifyContent": "space-between",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "paddingHorizontal": 10,
              "paddingVertical": 10,
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            multiline={true}
            onChangeText={[Function]}
            onContentSizeChange={[Function]}
            placeholder="Write comments"
            style={
              Array [
                undefined,
                undefined,
                Object {
                  "height": undefined,
                },
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#2E88FF",
                "borderRadius": 50,
                "height": 40,
                "justifyContent": "center",
                "opacity": 1,
                "width": 40,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={false}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#fff",
                    "fontSize": 20,
                  },
                  undefined,
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</BVLinearGradient>
`;
